### PR TITLE
Added foldr of permutation of Commutative Monoid

### PR DESCRIPTION
**Pull request by guilhermehas**
My code just added this proof:

```agda
foldr-cmonoid : xs ↭ ys → foldr _∙_ ε xs ≈ foldr _∙_ ε ys
```

### CODE REVIEW

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda [hunk 1]
```diff
@@ -342,3 +342,21 @@ module _ {ℓ} {R : Rel A ℓ} (R? : Decidable R) where
...
+
+------------------------------------------------------------------------
+-- foldr of Commutative Monoid
+
+module _ {c ℓ} (cmonoid : CommutativeMonoid c ℓ) where
+  open module CM = CommutativeMonoid cmonoid
+  import Relation.Binary.Reasoning.Setoid setoid as S
+  open import Algebra.Solver.CommutativeMonoid cmonoid
+
+  foldr-cmonoid : ∀ {xs ys} → xs ↭ ys → foldr _∙_ ε xs ≈ foldr _∙_ ε ys
```
Comment (MatthewDaggitt) [inline src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda:354]: Can we make this name more standard by changing it to `foldr-commMonoid`?

File: CHANGELOG.md [hunk 2]
```diff
@@ -3027,3 +3027,8 @@ This is a full list of proofs that have changed form to use irrelevant instance
   ```agda
   <-weakInduction-startingFrom : P i →  (∀ j → P (inject₁ j) → P (suc j)) → ∀ {j} → j ≥ i → P j
   ```
+
+* In `Data.List.Relation.Binary.Permutation.Propositional.Properties`:
+  ```agda
+  foldr-cmonoid : ∀ {xs ys} → xs ↭ ys → foldr _∙_ ε xs ≈ foldr _∙_ ε ys
```
Comment (MatthewDaggitt) [inline CHANGELOG.md:3033]: We can omit the implicit parameters in the entry. However we do need the `CommutativeMonoid` argument here.

**Review by MatthewDaggitt — COMMENTED** [review]
Thanks for the PR! A few comments.

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda [hunk 3]
```diff
@@ -342,3 +342,21 @@ module _ {ℓ} {R : Rel A ℓ} (R? : Decidable R) where
...
     x ∷ xs ++ y ∷ ys         ∎
     where open PermutationReasoning
+
+------------------------------------------------------------------------
+-- foldr of Commutative Monoid
+
+module _ {c ℓ} (cmonoid : CommutativeMonoid c ℓ) where
+  open module CM = CommutativeMonoid cmonoid
+  import Relation.Binary.Reasoning.Setoid setoid as S
+  open import Algebra.Solver.CommutativeMonoid cmonoid
```
[the remark this answers is not in the document]
Comment (guilhermehas) [reply src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda]: I removed the import.

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda [hunk 4]
```diff
@@ -342,3 +342,21 @@ module _ {ℓ} {R : Rel A ℓ} (R? : Decidable R) where
...
+
+  foldr-cmonoid : ∀ {xs ys} → xs ↭ ys → foldr _∙_ ε xs ≈ foldr _∙_ ε ys
+  foldr-cmonoid _↭_.refl = CM.refl
+  foldr-cmonoid (prep x xs↭xs) = ∙-congˡ (foldr-cmonoid xs↭xs)
+  foldr-cmonoid (swap {xs} {ys} x y xs↭ys) = S.begin
+    (x ∙ (y ∙ foldr _∙_ ε xs)) S.≈⟨ ∙-congˡ (∙-congˡ (foldr-cmonoid xs↭ys)) ⟩
+    (x ∙ (y ∙ foldr _∙_ ε ys))
+      S.≈⟨ solve 3 (λ x y ys → (x ⊕ y ⊕ ys) , (y ⊕ x ⊕ ys)) CM.refl x y (foldr _∙_ ε ys) ⟩
```
[the remark this answers is not in the document]
Comment (guilhermehas) [reply src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda]: I removed the solver.

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda [hunk 5]
```diff
@@ -342,3 +342,21 @@ module _ {ℓ} {R : Rel A ℓ} (R? : Decidable R) where
...
+module _ {c ℓ} (cmonoid : CommutativeMonoid c ℓ) where
+  open module CM = CommutativeMonoid cmonoid
+  import Relation.Binary.Reasoning.Setoid setoid as S
+  open import Algebra.Solver.CommutativeMonoid cmonoid
+
+  foldr-cmonoid : ∀ {xs ys} → xs ↭ ys → foldr _∙_ ε xs ≈ foldr _∙_ ε ys
+  foldr-cmonoid _↭_.refl = CM.refl
+  foldr-cmonoid (prep x xs↭xs) = ∙-congˡ (foldr-cmonoid xs↭xs)
+  foldr-cmonoid (swap {xs} {ys} x y xs↭ys) = S.begin
+    (x ∙ (y ∙ foldr _∙_ ε xs)) S.≈⟨ ∙-congˡ (∙-congˡ (foldr-cmonoid xs↭ys)) ⟩
```
[the remark this answers is not in the document]
Comment (guilhermehas) [reply src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda]: I removed the parenthesis.

File: src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda (hunk 1 above)
Comment (guilhermehas) [reply src/Data/List/Relation/Binary/Permutation/Propositional/Properties.agda]: I renamed it to foldr-commMonoid.

File: CHANGELOG.md (hunk 2 above)
Comment (guilhermehas) [reply CHANGELOG.md]: I removed ∀ {xs ys}

**Comment by gallais** [comment]
Should this go `Data.List.Relation.Binary.Permutation.Setoid.Properties` instead?
I am not sure there's anything specifically propositional about it?

**Comment by guilhermehas** [comment]
I moved from Propositional to Setoid.

**Review by gallais — APPROVED** [review]

File: src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda [hunk 6]
```diff
@@ -473,3 +475,34 @@ module _ {ℓ} {R : Rel A ℓ} (R? : B.Decidable R) where
...
+    x ∙ (y ∙ fεxs)   S.≈⟨ ∙-congˡ (∙-congˡ (foldr-commMonoid xs↭ys)) ⟩
+    x ∙ (y ∙ fεys)  S.≈˘⟨ assoc x y fεys ⟩
+    (x ∙ y) ∙ fεys   S.≈⟨ ∙-congʳ (comm x y) ⟩
+    (y ∙ x) ∙ fεys   S.≈⟨ ∙-congʳ (∙-cong y≈y′ x≈x′) ⟩
+    (y′ ∙ x′) ∙ fεys S.≈⟨ assoc y′ x′ fεys ⟩
+    y′ ∙ (x′ ∙ fεys) S.∎
+    where
+    fε = foldr _∙_ ε
+    fεxs = fε xs
+    fεys = fε ys
```
Comment (Taneb) [inline src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda:507]: These local definitions aren't worth it. `fε` isn't being used, and the other two only save you one character each. Please in-line and remove them.

**Review by Taneb — CHANGES_REQUESTED** [review]

File: src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda (hunk 6 above)
Comment (guilhermehas) [reply src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda]: I removed all of these definitions.

File: src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda [hunk 7]
```diff
@@ -473,3 +475,30 @@ module _ {ℓ} {R : Rel A ℓ} (R? : B.Decidable R) where
...
+
+  open ACM cmonoid
+
+  foldr-commMonoid : ∀ {xs ys} → xs ↭ ys → foldr _∙_ ε xs ≈ foldr _∙_ ε ys
+  foldr-commMonoid (refl []) = CM.refl
+  foldr-commMonoid (refl (x≈y ∷ xs≈ys)) = ∙-cong x≈y (foldr-commMonoid (Permutation.refl xs≈ys))
+  foldr-commMonoid (prep x≈y xs↭ys) = ∙-cong x≈y (foldr-commMonoid xs↭ys)
+  foldr-commMonoid (swap {xs} {ys} {x} {y} {x′} {y′} x≈x′ y≈y′ xs↭ys) = S.begin
+    x ∙ (y ∙ foldr _∙_ ε xs)   S.≈⟨ ∙-congˡ (∙-congˡ (foldr-commMonoid xs↭ys)) ⟩
+    x ∙ (y ∙ foldr _∙_ ε ys)  S.≈˘⟨ assoc x y (foldr _∙_ ε ys) ⟩
```
Comment (Taneb) [inline src/Data/List/Relation/Binary/Permutation/Setoid/Properties.agda:499]:
```suggestion
    x ∙ (y ∙ foldr _∙_ ε ys)   S.≈˘⟨ assoc x y (foldr _∙_ ε ys) ⟩
```

We generally align the `≈`, not the `⟨` when using equational reasoning

**Review by MatthewDaggitt — APPROVED** [review]

**Pull request merged**